text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,372 @@
+    - `fix-version-bump`: This would happen when a version bump was performed, but the value it was bumped to is incorrect.
+
+5. If the state of step 3 is `pass` or `pass-on-error`, we are done. Otherwise, *GOTO* 1.
+
+<br>
+
+<!-- To edit: URL Decode -> Edit -> URL Encode -->
+
+<img src='https://g.gravizo.com/svg?digraph%20G%20%7B%0A%20%20%20%20aize%20%3D%224%2C4%22%3B%0A%20%20%20%20%22merge-to-master%22%3B%0A%20%20%20%20%22merge-to-master%22%20-%3E%20%22test-framework-compatibility.sh%22%3B%0A%20%20%20%20node%20%5Bshape%3Dbox%2Cstyle%3Dfilled%2Ccolor%3D%22yellowgreen%22%5D%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20pass%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20%22pass-on-error%22%3B%0A%20%20%20%20node%20%5Bshape%3Dbox%2Cstyle%3Dfilled%2Ccolor%3D%22crimson%22%5D%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20%22fail%22%3B%0A%20%20%20%20%22test-framework-compatibility.sh%22%20-%3E%20%22fail-on-error%22%3B%0A%20%20%20%20%22fail%22%20-%3E%20%22merge-to-master%22%20%5Blabel%20%3D%20%22%20%20make-compatible%5CnOR%5Cnmark-incompatible%22%5D%0A%20%20%20%20%22fail-on-error%22%20-%3E%20%22merge-to-master%22%20%5Blabel%20%3D%20%22%20%20fix-version-bump%22%5D%0A%20%20%7D%0A'/>
+
+<br>
+
+# Drawbacks
+
+- Dramatically increases the duration of our integration tests suite.
+- Requires manually editing the framework compatibility test file.
+
+    > This isn't necessarily a drawback, as it provides a clear paper trail of all our breakages.
+
+# Rationale and Alternatives
+
+An alternative approach would to be force CLI versions to *always* be `>=` than Framework Versions. This differs from our current statement where we do try to have newer frameworks support older CLI as much as possible.
+
+Doing so drastically simplify the entire process, as it means we can effectively drop the entire notion of *cx-protocol*.
+
+## Why is it OK to enforce this requirement?
+
+- We want to migrate most of the functionality to the framework. Making CLI upgrades especially safe.
+- Users tend to upgrade the CLI before upgrading the framework.
+- We sometimes require that even now, and we haven't seen any backlash due to that.
+
+## Why is it NOT OK to enforce this requirement?
+
+- It is a bit intrusive. Even though we currently sometimes do it, we usually don't. And this would mean users get a slightly less smooth upgrade experience.
+
+## How does this affect the design?
+
+As far at testing strategy, it only affects the `test-framework-compatibility.sh` file, which would now look like this:
+
+```bash
+# Framework doesn't support older versions of the CLI, so we expect everything to fail with an appropriate error message.
+run_integration_tests_perm ${LATEST_PUBLISHED_VERSION} "SOURCE ${LATEST_PUBLISHED_VERSION} ${EXPECTED_ERROR}
```

Comment:
@eladb I know you mentioned in our discussion that this can probably just be a unit test. Though im not sure its that simple, compatibility testing cannot really be tested between in unit tests as it always uses the same versions.

In any case, this can be revised.